pkg/network/payload/notary_request.go

```diff
@@ -0,0 +1,144 @@
+	Network             netmode.Magic
+
+	Witness transaction.Witness
+
+	hash       util.Uint256
+	signedHash util.Uint256
+}
+
+// Hash returns payload's hash.
+func (r *P2PNotaryRequest) Hash() util.Uint256 {
+	if r.hash.Equals(util.Uint256{}) {
+		if r.createHash() != nil {
+			panic("failed to compute hash!")
+		}
+	}
+	return r.hash
+}
+
+// GetSignedHash returns a hash of the payload used to verify it.
+func (r *P2PNotaryRequest) GetSignedHash() util.Uint256 {
+	if r.signedHash.Equals(util.Uint256{}) {
+		if r.createHash() != nil {
+			panic("failed to compute hash!")
+		}
+	}
+	return r.signedHash
+}
+
+// GetSignedPart returns a part of the payload which must be signed.
+func (r *P2PNotaryRequest) GetSignedPart() []byte {
+	buf := io.NewBufBinWriter()
+	buf.WriteU32LE(uint32(r.Network))
+	r.encodeHashableFields(buf.BinWriter)
+	if buf.Err != nil {
+		return nil
+	}
+	return buf.Bytes()
+}
+
+// createHash creates hash of the payload.
+func (r *P2PNotaryRequest) createHash() error {
+	b := r.GetSignedPart()
+	if b == nil {
+		return errors.New("failed to serialize hashable data")
+	}
+	r.updateHashes(b)
+	return nil
+}
+
+// updateHashes updates Payload's hashes based on the given buffer which should
+// be a signable data slice.
+func (r *P2PNotaryRequest) updateHashes(b []byte) {
+	r.signedHash = hash.Sha256(b)
+	r.hash = hash.Sha256(r.signedHash.BytesBE())
+}
+
+// DecodeBinaryUnsigned reads payload from w excluding signature.
+func (r *P2PNotaryRequest) decodeHashableFields(br *io.BinReader) {
+	r.MainTransaction = new(transaction.Transaction)
+	r.FallbackTransaction = new(transaction.Transaction)
+	r.MainTransaction.DecodeBinary(br)
+	r.FallbackTransaction.DecodeBinary(br)
+	if br.Err == nil {
+		br.Err = r.isValid()
+	}
+	if br.Err == nil {
+		br.Err = r.createHash()
+	}
+}
+
+// DecodeBinary implements io.Serializable interface.
+func (r *P2PNotaryRequest) DecodeBinary(br *io.BinReader) {
+	r.decodeHashableFields(br)
+	if br.Err == nil {
+		r.Witness.DecodeBinary(br)
+	}
+}
+
+// encodeHashableFields writes payload to w excluding signature.
+func (r *P2PNotaryRequest) encodeHashableFields(bw *io.BinWriter) {
+	r.MainTransaction.EncodeBinary(bw)
+	r.FallbackTransaction.EncodeBinary(bw)
+}
+
+// EncodeBinary implements Serializable interface.
+func (r *P2PNotaryRequest) EncodeBinary(bw *io.BinWriter) {
+	r.encodeHashableFields(bw)
+	r.Witness.EncodeBinary(bw)
+}
+
+func (r *P2PNotaryRequest) isValid() error {
+	nKeysMain := r.MainTransaction.GetAttributes(transaction.NotaryAssistedT)
+	if len(nKeysMain) == 0 {
+		return errors.New("main transaction should have NotaryAssisted attribute")
+	}
+	if nKeysMain[0].Value.(*transaction.NotaryAssisted).NKeys == 0 {
+		return errors.New("main transaction should have NKeys > 0")
+	}
+	// TODO: check "either doesn't have all witnesses attached (in this case none of them can be multisignature), or it only has a partial multisignature, only one of the two is allowed "
```

Comment:
Maybe we can leave it to the module. The payload as it is now is guaranteed to have valid fallback transaction and random node on the network should probably only care about that.